lib/template-loader.php

```diff
@@ -223,6 +234,10 @@ function gutenberg_find_template_post_and_parts( $template_hierarchy ) {
 
 	$current_template_post = $template_query->have_posts() ? $template_query->next_post() : null;
 
+	// TODO: We could consider moving the following logic to a hook like `posts_results` that is run
+	// after query results have been fetched. That might allow us to absorb filtering logic
+	// (as e.g. found in `filter_rest_wp_template_query`) into the generic querying logic.
```

Comment:
This comment is confusing. How would moving this into a hook allow for `filter_rest_wp_template_query` to be in the "generic querying logic."?

Reply:
What I meant is to move the auto-draft creation logic into a `post_results` filter, and to filter the results returned by the query to only return the template(s) that are relevant for the requested template hierarchy -- similar to what `filter_rest_wp_template_query` does in this PR.

Reply:
The auto-drafts are not created for the templates that are returned by the query. They are created for spots in the template hierarchy that don't have a returned template but have a theme file. You would have to create the auto-drafts before making the query so that they show up, but give preference to non-auto-drafts. That could work, but I doubt it will be more straightforward.

Reply:
Yeah, I think we're basically on the same page. <y thinking was basically:

- We tap into the `post_results` hook to watch out for a query for one (or many) `wp_template`(s), possibly with a `post_name` or `post_name__in` param, akin to this: https://github.com/WordPress/gutenberg/blob/8676302f5cc27838d0c7eb586d2384041ece0ed2/lib/template-loader.php#L212-L222
- In our filter, we create the auto-drafts from theme files, and if there's one that matches the template hierarchy better than our query result, we return that -- that's basically https://github.com/WordPress/gutenberg/blob/8676302f5cc27838d0c7eb586d2384041ece0ed2/lib/template-loader.php#L224-L280

While it mangles the query logic quite a bit, it's arguable that we'd pretty much always like to perform these steps whenever a template is queried (whether for rendering or through the REST API). I think it might be worth experimenting with.

Reply:
We could also just call the new functions directly for rendering.